### PR TITLE
 Fix #6197: erase pattern bound type symbols completely

### LINKED ISSUE #6197: covariant type parameter in sealed trait causes match may not be exhaustive
```scala
object Test {
  sealed trait Cause[+E]

  object Cause {
    final case class Fail[E](value: E) extends Cause[E]
  }

  def fn(cause: Cause[Any]): String =
    cause match {
      case Cause.Fail(t: Throwable) => t.toString
      case Cause.Fail(any)          => any.toString
    }
}
```

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #6197: erase pattern bound type symbols completely

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -376,13 +365,18 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
...
   /* Erase pattern bound types with WildcardType */
-  def erase(tp: Type): Type = {
+  def erase(tp: Type): Type = trace(i"$tp erased to", debug) {
     def isPatternTypeSymbol(sym: Symbol) = !sym.isClass && sym.is(Case)
 
     val map = new TypeMap {
       def apply(tp: Type) = tp match {
+        case tp @ AppliedType(tycon, args) if (tycon.isRef(defn.ArrayClass)) =>
+          // cannot use WildcardType for Array[_], due to that
+          //   Array[WildcardType] <: Array[Array[WildcardType]]
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: But this isn't specific to Array, the same should be true for other type constructors, so I don't understand why Array in particular is special-cased.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: A difference is that `_: Array[Array[_]]`  does not make `_: Array[_]` unreachable, but it's true for class type constructors.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Aha, I see, thanks! Could you expand the comment a bit to make this more clear ? Also it'd be nice if the documentation of `erase` contained some more details/examples explaining why erasing pattern bound types with WildcardType is a good idea. I'm also wondering what happens with type constructors which are not classes (type lambdas, match types), are they handled somewhere else ?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Thanks @smarter. I just add a document with examples. Type lambdas used for type test will result in uncheckable at runtime warnings. We haven't seen match types used in patterns, currently it just keeps the type as it is.

**Pull request merged**